CA: DRA integration MVP

towca opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

This is a part of Dynamic Resource Allocation (DRA) support in Cluster Autoscaler.

This PR implements an MVP of DRA integration in Cluster Autoscaler. Not all CA features work with DRA yet (see list below), but most logic is implemented and DRA autoscaling can be tested in a real cluster.

Changes summary:

* Implement a bunch of DRA-related utils. Ideally these would be upstream, but IMO we can revisit this later.
* Introduce `dynamicresources.Provider` which retrieves and snapshots all DRA objects.
* Introduce `dynamicresources.Snapshot` which allows modifying the DRA objects snapshot obtained from `Provider`, and exposes the DRA objects to the scheduler framework.
* Implement a very rudimentary utilization calculation for device pools, so that scale-down has something to act on.
* Start sanitizing the DRA objects when duplicating Nodes and their DS-like pods in the snapshot.
* Modify DRA objects in the `dynamicresources.Snapshot` saved in `SnapshotBase` during `ClusterSnapshot` mutations in `PredicateChecker`.
* Add `StaticAutoscaler` integration-like unit tests covering DRA scenarios.

The following features don't work with DRA yet, and will be tackled post-MVP:

* **Priority-based preempting pods using DRA.** If CA sees an unschedulable pod waiting for scheduler preemption (with `nominatedNodeName` set), it adds the pod to the `nominatedNodeName` in the snapshot without checking predicates, or even removing the preempted pod (so the Node can be effectively "overscheduled"). If such a Pod uses DRA, we'll have to run scheduler predicates to actually obtain the necessary ResourceClaim allocations. If we just force-add such a Pod to the snapshot without modifying the claims, CA doesn't see the Node's ResourceSlices as used and can just schedule another Pod to use them in the simulations.
* **DaemonSet/static pods using DRA (in some cases).** Similarly to the first point, DS/static pods using DRA won't work correctly during scale-from-0-nodes scenarios where `TemplateNodeInfo()` has to be called (since scheduler predicates aren't currently run there). Forcing "missing" DS pods onto template Nodes also won't work if the DS pods use DRA (we'll have to duplicate their resource claims and start running scheduler predicates there as well).
* **Delta ClusterSnapshot base.** Only the `BasicSnapshotBase` (old `BasicClusterSnapshot`) implementation is integrated with `dynamicresources.Snapshot`. `DeltaSnapshotBase` (old `DeltaClusterSnapshot`) isn't yet. We'll need to add some delta-like capability to `dynamicresources.Snapshot` for that in addition to just `Clone()`.

Additionally, the following point will have to be tackled post-MVP:
* As pointed out by @MaciekPytel during the initial review of #7350, the state for unschedulable pods is kept in two separate places after this PR. The pod objects themselves are just a list obtained from listers in `RunOnce`, then processed by `PodListProcessors` and passed to `ScaleUp`. The Pods' DRA objects, on the other hand, live in the `dynamicresources.Snapshot` inside `ClusterSnapshot`. This leaves us with a risk of the two data sources diverging quite easily (e.g. a `PodListProcessor` injecting a "fake" unschedulable Pod to the list, but not injecting the Pod's ResourceClaims to the `ClusterSnapshot`). To make this better, in a follow-up PR the unschedulable pods will be fully moved inside `ClusterSnapshot` and `PodListProcessors` will interact with them via `ClusterSnapshot` methods instead of getting them by argument and returning.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/118612

#### Special notes for your reviewer:

The first commit in the PR is just a squash of #7479, #7497, and #7529, and it shouldn't be a part of this review. The PR will be rebased on top of master after the others are merged.

#### Does this PR introduce a user-facing change?

```release-note
TODO
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
https://github.com/kubernetes/enhancements/blob/9de7f62e16fc5c1ea3bd40689487c9edc7fa5057/keps/sig-node/4381-dra-structured-parameters/README.md
```

Shubham82 commented:
Hi @towca, the `Tests / test-and-verify` test case failed, which is required to merge this PR.
the error is related to the boilerplate. Here is the error:

`The boilerplate header is wrong for /cluster-autoscaler/simulator/dynamicresources/utils/utilization.go`

Shubham82 commented on cluster-autoscaler/simulator/dynamicresources/utils/utilization.go:
```diff
@@ -0,0 +1,75 @@
+package utils
```
before this line, you have to add the following comment:

```
/*
Copyright 2020 The Kubernetes Authors.
Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at
    http://www.apache.org/licenses/LICENSE-2.0
Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/
```

Shubham82 commented:
> The first commit in the PR is just a squash of https://github.com/kubernetes/autoscaler/pull/7479, https://github.com/kubernetes/autoscaler/pull/7497, and https://github.com/kubernetes/autoscaler/pull/7529, and it shouldn't be a part of this review. The PR will be rebased on top of master after the others are merged.


As mentioned in the PR description, I am putting this PR on hold to avoid an accidental merge.

/hold

pohly commented:
/wg device-management

pohly commented:
I only looked at the PR description so far. I am not sure whether I am qualified to review the code :cold_sweat: 

> Implement some utils for interacting with ResourceClaims. These would probably ideally be upstream, but IMO we can revisit this later.

Agreed, this can come later. https://pkg.go.dev/k8s.io/dynamic-resource-allocation@v0.31.4/resourceclaim might be the right package for that. If "resourceclaim" works as a package name, then you could already use it now and then later only the import path needs to be changed once that new code is upstream.

One thing that I don't see mentioned is node readiness: directly after provisioning a new node, the DRA driver pod(s) have to be started on the node and publish their ResourceSlices before the new node is really ready to support pods using claims. I know that there are checks for device plugins to treat such a node as "not ready" while plugins are starting. Is there something similar for DRA drivers? This might not be needed for MVP.

towca commented:
> I only looked at the PR description so far. I am not sure whether I am qualified to review the code 😰
> 
> >

Yeah unfortunately it's quite a lot, IMO we should split reviewing responsibilities here:
* @BigDarkClown you're probably the most familiar with core CA code, could you focus your review on how the DRA logic fits into the existing CA logic? In particular:
  * Could you verify that the behavior doesn't change if the DRA flag is disabled?
  * Could you review the error handling?
  * Could you check if I missed any places that need to take DRA into account (beside the parts with TODO(DRA))?
  * Could you review the code structure etc.?
* @pohly you're definitely the most familiar with the DRA logic/assumptions, could you focus your review on the DRA parts and if the assumptions match? In particular:
  * ResourceClaim utils (1st commit) - is the logic correct there?
  * "Sanitizing" ResourceClaims (4th commit). During the simulations, CA duplicates Nodes and their DaemonSet pods. Duplicated Nodes/Pods go through "sanitization" so that they aren't actually identical to the originals. Doing this for DS pods using DRA has some corner cases (described in the comments), WDYT about this? Do we anticipate DS pods using DRA?
  * Calculating utilization (5th commit + the PR description) - does my initial "provisional" idea make sense? Is it implemented correctly?
  * Modifying the state of DRA objects during scheduling simulations (6th commit). This is tied to the ResourceClaim utils, as a lot of them are used here. Is the logic correct? One question that comes to mind - if we remove the last pod reserving a shared claim, should it be deallocated?
  * Integration tests (7th commit) - do the scenarios listed there make sense and behave as you would expect them to? Are some major scenarios missing?
* @jackfrancis AFAIU you're familiar with both the DRA and CA parts, could you focus your review somewhere "between" the two parts above?

/assign @BigDarkClown 
/assign @pohly 
/assign @jackfrancis

> One thing that I don't see mentioned is node readiness: directly after provisioning a new node, the DRA driver pod(s) have to be started on the node and publish their ResourceSlices before the new node is really ready to support pods using claims. I know that there are checks for device plugins to treat such a node as "not ready" while plugins are starting. Is there something similar for DRA drivers? This might not be needed for MVP.

I was wondering about that part, hard to test it without an actual driver (the test driver from k/k e2e tests doesn't run into this). So a new Node can be Ready but not yet have its ResourceSlices exposed, is that right? If so then we indeed need to reimplement the device plugin logic. How can CA know that a Node is supposed to have any slices exposed though? For device plugins we depend on a specific label on the Node (if the label is on the Node but no allocatable resource -> fake the node as not ready).

@jackfrancis could you verify how this behaves in practice during your manual testing? Are we seeing double scale-ups?

jackfrancis commented:
@pohly @towca in my testing (A100 nodes using the nvidia [k8s-dra-driver](https://github.com/NVIDIA/k8s-dra-driver)) I didn't see any undesirable functional side-effects of the node bootstrapping process. It's definitely true that before a pending pod+resourceclaim can be fulfilled there's a lot more waiting than normal:

1) new node comes online
2) dra driver daemonset pods land on new node and perform host OS level driver installation
3) gpu operator carries out mig device configuration
4) node is rebooted
5) resourceslice is populated

During that period of time, I observed that cluster-autoscaler sanely (1) notices that the new node is there but that (2) the resourceclaim is not yet able to be fulfilled. I didn't notice any TTL expire like "I'm tired of waiting for this new node, I'm going to try and build a new one".

In practice, my experiments produced a ~10 minute wait between node provisioning and resourceslice population. That's a long time, admittedly! Is there a particular configuration we could set to try to reproduce any "double scale out" symptoms in my environment?

For full disclosure, this is the workload I used to test:

```yaml
apiVersion: resource.k8s.io/v1alpha3
kind: ResourceClaimTemplate
metadata:
  namespace: gpu-test
  name: mig-devices
spec:
  spec:
    devices:
      requests:
      - name: mig-1g-10gb
        deviceClassName: mig.nvidia.com
        selectors:
        - cel:
            expression: "device.attributes['gpu.nvidia.com'].profile == '1g.10gb'"
---
apiVersion: apps/v1
kind: Deployment
metadata:
  namespace: gpu-test
  name: gpu-test
  labels:
    app: gpu-test-pod
spec:
  replicas: 1
  selector:
    matchLabels:
      app: gpu-test
  template:
    metadata:
      labels:
        app: gpu-test
    spec:
      runtimeClassName: nvidia
      resourceClaims:
      - name: mig-device-gpu-test
        resourceClaimTemplateName: mig-devices
      containers:
      - name: mig-device-container
        image: ubuntu:22.04
        command: ["bash", "-c"]
        args: ["nvidia-smi -L; trap 'exit 0' TERM; sleep 9999 & wait"]
        resources:
          claims:
          - name: mig-device-gpu-test
            request: mig-1g-10gb
```

1) create cluster w/ 1 A100 node that exposes 7 * 10GB mig devices

```
      nvidia.com/mig.capable: "true"
      nvidia.com/mig.config: all-1g.10gb
      nvidia.com/mig.config.state: success
```

2) scale deployment replicas out to 8, which creates 1 pending pod + resourceclaim
3) wait for CA to add new node and observe (eventually) all 8 pods are running + all resourceclaims are allocated,reserved

pohly commented:
> How can CA know that a Node is supposed to have any slices exposed though?

Isn't that what needs to be configured for a node pool? Without information about what devices are going to be made available together with a new node, CA cannot scale up. This readiness check could compare that assumption against the actual devices.

> During that period of time, I observed that cluster-autoscaler sanely (1) notices that the new node is there but that (2) the resourceclaim is not yet able to be fulfilled.

I don't remember what the exact problem is, i.e. whether it is "not scaling up when it should" or "scaling down when it shouldn't". I guess that it works in practice indicates that it is not needed for a MVP.

pohly commented on cluster-autoscaler/simulator/dynamicresources/utils/resource_claims.go:
```diff
@@ -0,0 +1,134 @@
+		return
+	}
+
+	newReservedFor := make([]resourceapi.ResourceClaimConsumerReference, 0, len(claim.Status.ReservedFor))
```
Nit: I would find the right index, then use `slices.Delete` to create the new slice minus that one entry.

pohly commented on cluster-autoscaler/simulator/dynamicresources/utils/resource_claims_test.go:
```diff
@@ -0,0 +1,735 @@
+func TestClaimOwningPod(t *testing.T) {
+	truePtr := true
+	for _, tc := range []struct {
+		testName string
```
Personally I prefer `map[]struct` because it ensures at compile time that I do not reuse the same test name twice.

Also, better avoid spaces in test names. They get replaced with underscores by `go test` and then the tests listed in the output can't be found easily in the source code. `go test -run=...` also is more complicated to get right.

pohly replied:
> benefits of descriptive test names outweigh the downsides

They can still be as descriptive as they are now, just don't use spaces as separators. If it's not readable without them, then it's also not readable in the `go test` output.

> I'd leave it for consistency

Consistency certainly trumps my personal preferences, :+1:

pohly reviewed (commented):
> ResourceClaim utils (1st commit) - is the logic correct there?

Logic is correct. I would do a few things differently myself, but that is not a blocker for the PR.

BigDarkClown reviewed (commented):
Left some comments, I am not the expert on DRA so I might have missed something. Overall this LGTM for the CA side, but please get an additional review for DRA handling.

pohly commented on cluster-autoscaler/simulator/node_info_utils.go:
```diff
@@ -41,6 +42,9 @@ type nodeGroupTemplateNodeInfoGetter interface {
 func SanitizedTemplateNodeInfoFromNodeGroup(nodeGroup nodeGroupTemplateNodeInfoGetter, daemonsets []*appsv1.DaemonSet, taintConfig taints.TaintConfig) (*framework.NodeInfo, errors.AutoscalerError) {
+	// TODO(DRA): Figure out how to handle TemplateNodeInfo() returning DaemonSet Pods using DRA. Currently, things only work correctly if such pods are
+	// already allocated by TemplateNodeInfo(). It might be better for TemplateNodeInfo() to return unallocated claims, and to run scheduler predicates and
+	// compute the allocations here.
```
https://github.com/kubernetes/autoscaler/pull/7530/#issuecomment-2536807564:
> "Sanitizing" ResourceClaims (4th commit). During the simulations, CA duplicates Nodes and their DaemonSet pods. Duplicated Nodes/Pods go through "sanitization" so that they aren't actually identical to the originals. Doing this for DS pods using DRA has some corner cases (described in the comments), WDYT about this? Do we anticipate DS pods using DRA?

A use case for a DS with claims is for monitoring, in which the claims will use "admin mode", meaning that they don't block allocating the monitored devices for some other pod. Admin mode is under a feature gate in Kubernetes and alpha, so supporting this in CA is not a priority.

pohly commented on cluster-autoscaler/simulator/node_info_utils.go:
```diff
@@ -58,33 +62,44 @@ func SanitizedTemplateNodeInfoFromNodeInfo(example *framework.NodeInfo, nodeGrou
 	for _, podInfo := range nodeInfo.Pods() {
 		freshPod := sanitizePod(podInfo.Pod, freshNode.Name, namesSuffix)
-		result.AddPod(framework.NewPodInfo(freshPod, nil))
+		freshResourceClaims, err := drautils.SanitizePodResourceClaims(freshPod, podInfo.Pod, podInfo.NeededResourceClaims, namesSuffix, nodeInfo.Node().Name, freshNodeName, oldPoolNames)
```
It's not immediately obvious that `drautils.SanitizePodResourceClaims` doesn't change something in place. Perhaps rename it to `drautils.SanitizePodResourceClaimsDeepCopy` for consistency with this `NodeInfoSanitizedDeepCopy` here?

Or is it a convention that `sanitize<something>` returns a deep copy? `sanitizePod` and `sanitizeNode` below do that. Then why the `DeepCopy` suffix in `NodeInfoSanitizedDeepCopy`?

jackfrancis replied:
+1, it seems like all of the `sanitize.go` funcs produce a new copy. Would the following names be clearer?

- `GetSanitizedNodeResourceSlices`
- `GetSanitizedPodResourceClaims`
- `GetSanitizedResourceClaimRefs`

Or maybe follow the pattern from this package, e.g. `SanitizedTemplateNodeInfoFromNodeGroup` (note the "d" in Sanitized...) I probably prefer the clarity of the `Get` prefix but I understand why that might not be needed.

And probably `NodeInfoSanitizedDeepCopy` can be changed to either `SanitizedNodeInfo` or `GetSanitizedNodeInfo` (depending on what pattern we choose to standardize to).

towca replied:
Good catch, it's indeed a bit of convention that the `Sanitized<something>` functions return a deep-copy, but as Jack noted the "d" is important here - my bad. Standardized the names following Jacks suggestions:

* `SanitizeNodeResourceSlices` -> `SanitizedNodeResourceSlices` (without the Get prefix to match Go conventions, as it's an exported function)
* `SanitizePodResourceClaims` -> `SanitizedPodResourceClaims` (same reasoning as above)
* `SanitizeResourceClaimRefs` -> `SanitizedResourceClaimRefs` (same reasoning as above)
* `NodeInfoSanitizedDeepCopy` -> `SanitizedNodeInfo` (same reasoning as above)
* `sanitizeNodeInfo` -> `createSanitizedNodeInfo` (added the "create" prefix here as it's a local function and `sanitizedNodeInfo` is used as a variable name, "create" fits better than "get" IMO).
* `sanitizeNode` -> `createSanitizedNode` (same reasoning as above).
* `sanitizePod` -> `createSanitizedPod` (same reasoning as above).

pohly commented on cluster-autoscaler/simulator/node_info_utils.go:
```diff
@@ -130,6 +145,9 @@ func podsExpectedOnFreshNode(sanitizedExampleNodeInfo *framework.NodeInfo, daemo
 	// Add all pending DS pods if force scheduling DS
+	// TODO(DRA): Figure out how to make this work for DS pods using DRA. Currently such pods would get force-added to the
+	// ClusterSnapshot, but the ResourceClaims reflecting their DRA usage on the Node wouldn't. So CA would be overestimating
+	// available DRA resources on the Node.
```
This should be okay in practice for the "admin mode" monitoring use case because all devices on the new node would be available to pods, despite also having such a DS.

pohly reviewed (commented):
Reviewed some parts of " CA: sanitize and propagate DRA objects through NodeInfos in node_info utils".

pohly commented on cluster-autoscaler/simulator/dynamicresources/utils/utilization.go:
```diff
@@ -0,0 +1,163 @@
+			poolDevices := getAllDevices(currentSlices)
+			allocatedDeviceNames := allocatedDevices[driverName][poolName]
+			unallocated, allocated := splitDevicesByAllocation(poolDevices, allocatedDeviceNames)
+			result[driverName][poolName] = calculatePoolUtil(unallocated, allocated)
```
https://github.com/kubernetes/autoscaler/pull/7530/#issuecomment-2536807564:
> Calculating utilization (5th commit + the PR description) - does my initial "provisional" idea make sense? Is it implemented correctly?

This looks good to me for now.

It might have to become a bit more complex once partitionable devices get added (targeting 1.33 with that) because then a single GPU will be represented by different "logical" devices in a slice and the "size" of of those matters. For example, on one node one device representing an entire GPU might be allocated (one device out of many allocated) while on some other node, multiple smaller subsets of a similar GPU might be allocated (several devices out of many allocated). Then simply counting "allocated devices" would give the second node a higher utilization although the GPU is less utilized.

I'm not sure yet how this can be handled in a generic way. Perhaps we need to standardize a special attribute of devices which names a quantity that can serve as stand-in for the size of a device? For GPUs, that could be the memory size.

/cc @klueska @mortent @johnbelamaric

pohly commented on cluster-autoscaler/simulator/clustersnapshot/predicate/predicate_snapshot.go:
```diff
@@ -101,6 +170,30 @@ func (s *PredicateSnapshot) SchedulePodOnAnyNodeMatching(pod *apiv1.Pod, anyNode
+		if len(foundPod.Spec.ResourceClaims) > 0 {
+			if err := s.ClusterSnapshotStore.DraSnapshot().UnreservePodClaims(foundPod); err != nil {
+				return err
+			}
```
https://github.com/kubernetes/autoscaler/pull/7530/#issuecomment-2536807564
> Modifying the state of DRA objects during scheduling simulations (6th commit). This is tied to the ResourceClaim utils, as a lot of them are used here. Is the logic correct? One question that comes to mind - if we remove the last pod reserving a shared claim, should it be deallocated?

Yes, it should be treated as deallocated. Otherwise you cannot move such a pod from one node to another during a simulated scale down. It would be forced to stay on the original node because that is where the claim is allocated.

The rest of the logic looks okay to me.

pohly commented on cluster-autoscaler/core/static_autoscaler_dra_test.go:
```diff
@@ -0,0 +1,777 @@
+				{nodeName: node3GpuA1slice.name + "-1", reason: simulator.NotUnderutilized},
+				{nodeName: node3GpuA1slice.name + "-2", reason: simulator.NoPlaceToMovePods},
+			},
+		},
```
https://github.com/kubernetes/autoscaler/pull/7530/#issuecomment-2536807564
> Integration tests (7th commit) - do the scenarios listed there make sense and behave as you would expect them to? Are some major scenarios missing?

A scale down scenario where a pod using a claim gets moved successfully to some other node would be good. As discussed in https://github.com/kubernetes/autoscaler/pull/7530/#discussion_r1882417954, this probably doesn't work yet unless deallocation of the claim gets added. That scenario then also can cover utilization calculation (the "least utilized" node is meant to get removed).

A [monitoring DS with admin mode](https://github.com/kubernetes/autoscaler/pull/7530/#discussion_r1882368788) might also be useful, once support for that gets added.

Specific scenarios aside, a general comment about testing: I see that DRA support is behind a flag. Would it be possible to run all existing test scenarios with and without that flag? Nothing should change, but one cannot be sure unless it gets tested...

The inverse is also true: what happens in a cluster where DRA is used and the flag is off? The outcome is probably different from the "flag is on" case, but is it still sane?

towca replied:
> A scale down scenario where a pod using a claim gets moved successfully to some other node would be good. As discussed in #7530, this probably doesn't work yet unless deallocation of the claim gets added. That scenario then also can cover utilization calculation (the "least utilized" node is meant to get removed).

Doesn't "scale-down: single-device nodes with drain" cover that? Deallocating claims is already implemented, I was just wondering if it was the correct behavior to simulate.

> A monitoring DS with admin mode might also be useful, once support for that gets added.

Definitely, these tests should be expanded for the post-MVP features.

> Specific scenarios aside, a general comment about testing: I see that DRA support is behind a flag. Would it be possible to run all existing test scenarios with and without that flag? Nothing should change, but one cannot be sure unless it gets tested...
> 
> The inverse is also true: what happens in a cluster where DRA is used and the flag is off? The outcome is probably different from the "flag is on" case, but is it still sane?

* So all existing pre-DRA testing scenarios are run with the DRA flag enabled, to cover more code paths and ensure that previous functionality still works as expected with DRA enabled but no DRA objects to work on.
* I agree that ideally we'd also run all of them with DRA disabled and no DRA objects to work on for 100% certainty that the behavior stays the same, but that seems hard to do technically. `PredicateSnapshot()` takes the DRA flag, and it's used in unit tests throughout the codebase. Right now all these tests just call `testsnapshot.TestClusterSnapshotOrDie(t)` to get a test snapshot (with DRA enabled as described above), and I think they would all need to be modified to iterate over both flag options. This seems like a lot of work for slightly more test coverage.
* Now the "DRA is disabled but there are DRA objects" is something that's indeed not well covered. I don't think covering it is critical for MVP, added a TODO for post-MVP. WDYT?

pohly replied:
> Doesn't "scale-down: single-device nodes with drain" cover that? Deallocating claims is already implemented, I was just wondering if it was the correct behavior to simulate.

Then we are good.

> This seems like a lot of work for slightly more test coverage.

Ack.

> Now the "DRA is disabled but there are DRA objects" is something that's indeed not well covered. I don't think covering it is critical for MVP, added a TODO for post-MVP. WDYT?

I'm okay with extending code coverage later.

jackfrancis commented on cluster-autoscaler/core/scaledown/actuation/actuator.go:
```diff
@@ -368,7 +371,16 @@ func (a *Actuator) createSnapshot(nodes []*apiv1.Node) (clustersnapshot.ClusterS
+	if a.ctx.DynamicResourceAllocationEnabled && a.draProvider != nil {
+		draSnapshot, err = a.draProvider.Snapshot()
+		if err != nil {
+			// TODO(DRA): Maybe proceed?
```
I would vote to remove this TODO statement and stick with returning an error. Maybe we can add more strong typing to the error to make these unexpected DRA outcomes easier to discover in logs?

jackfrancis commented on cluster-autoscaler/core/scaledown/eligibility/eligibility_test.go:
```diff
@@ -69,51 +74,99 @@ func getTestCases(ignoreDaemonSetsUtilization bool, suffix string, now time.Time
 			nodes:            []*apiv1.Node{regularNode, justDeletedNode},
-			want:             []string{"regular"},
+			wantUnneeded:     []string{"regular"},
+			wantUnremovable:  []*simulator.UnremovableNode{{Node: justDeletedNode, Reason: simulator.CurrentlyBeingDeleted}},
```
+1 on improving the fidelity of existing tests

jackfrancis commented on cluster-autoscaler/core/static_autoscaler.go:
```diff
@@ -337,8 +341,17 @@ func (a *StaticAutoscaler) RunOnce(currentTime time.Time) caerrors.AutoscalerErr
+	if a.AutoscalingContext.DynamicResourceAllocationEnabled && a.draProvider != nil {
+		draSnapshot, err = a.draProvider.Snapshot()
+		if err != nil {
+			// TODO(DRA): Maybe proceed?
```
same comment as above about removing the TODO statement, and I'm not sure that `"apiCallError"` is the right error type here?

jackfrancis commented on cluster-autoscaler/main.go:
```diff
@@ -684,6 +692,15 @@ func main() {
+	// If the DRA flag is passed, we need to set the DRA feature gate as well. The selection of scheduler plugins for the default
+	// scheduling profile depends on feature gates, and the DRA plugin is only included if the DRA feature gate is enabled. The DRA
+	// plugin itself also checks the DRA feature gate and doesn't do anything if it's not enabled.
+	if *enableDynamicResourceAllocation && !featureGate.Enabled(features.DynamicResourceAllocation) {
```
I don't want to slow down integration but it's a bit of a shame we can't simply wire in the k/k feature gate interface in here for the user to specify natively in CAS (rather than manually setting it here based on CAS runtime config)

jackfrancis replied:
> CA already allows you to pass the feature gate flag

Do you mean inherit a feature gate from kubelet, kube-scheduler, etc, or use the feature gate facility instead of the command line config facility? I would prefer using a feature gate approach (user expresses a feature gate config to the CAS runtime) so that we could eventually graduate it (first set to true by default and then retire altogether like k/k does) and simplify the code here. I wouldn't want to have "if DRA then do this" code paths here forever.

towca replied:
> Do you mean inherit a feature gate from kubelet, kube-scheduler, etc, or use the feature gate facility instead of the command line config facility?

I don't think the feature gate flags are "centralized" or "inherited" between different components, it seems that they are set separately for each component (see the second sentence here: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#overview). The flags are likely synchronized between the components by something that deploys Kubernetes, but I'm not sure if that something would know to pass the flag to Cluster Autoscaler anyway.

I fully agree with not keeping the DRA flag forever in CA and removing it once we're confident in the implementation, but I think we can do that with a CA-specific flag as well. `--enable-dynamic-resource-allocation` is basically a lightweight, CA-specific feature gate - we've been following this pattern for previous features as well.

Especially because I'm pretty sure the graduation criteria are not the same here - DRA went to beta in 1.32, but DRA autoscaling is definitely an alpha in 1.32. If we were to just follow the graduation path for the DRA feature gate, we'd IMO be graduating things to early in CA. It seems that we'd have more control over the graduation if we use a dedicated flag.

jackfrancis replied:
Sounds like we can do this for the CA-specific `--enable-dynamic-resource-allocation` flag:

1) set default to true when DRA goes GA in k/k (for example, 1.34 or 1.35)
2) deprecate flag w/ a 4 release warning ("Warning, this flag will go away in 1.38")
3) remove flag altogether in (e.g.,) 1.38

Those are just example k8s versions, but something like that will allow us to simplify a lot of this code in the next few years, which will be beneficial to the project.

towca replied:
Yes, this is definitely the plan!

towca replied:
(hopefully earlier than 1.38 :sweat_smile:)

towca commented:
@BigDarkClown @jackfrancis @pohly Addressed all your comments! I don't think there is anything blocking this PR in the comments, I'd appreciate if we could close on it ASAP so that we can proceed with 1.32 CA release.

jackfrancis commented:
/test pull-cluster-autoscaler-e2e-azure-master

BigDarkClown commented on cluster-autoscaler/simulator/dynamicresources/snapshot/snapshot.go:
```diff
@@ -17,32 +17,244 @@ limitations under the License.
+	for _, podClaimRef := range pod.Spec.ResourceClaims {
+		claimName := claimRefToName(pod, podClaimRef)
+		if claimName == "" {
+			continue
```
Ok, I get the reasoning now. I would however add a comment here, to ensure that in the future we now explicitly why this error is not returned, and instead we proceed to the next claim.

towca replied:
Added a mention about the method not erroring out to the method comment, and comments to the continues. WDYT?

BigDarkClown commented on cluster-autoscaler/simulator/dynamicresources/snapshot/snapshot.go:
```diff
@@ -17,32 +17,244 @@ limitations under the License.
+		claimId := ResourceClaimId{Name: claimName, Namespace: pod.Namespace}
+		claim, found := s.resourceClaimsById[claimId]
+		if !found {
+			continue
```
Ditto, please add an explicit comment why we do that.

towca replied:
Done

BigDarkClown commented on cluster-autoscaler/simulator/dynamicresources/utils/sanitize.go:
```diff
@@ -0,0 +1,166 @@
+		newSlices = append(newSlices, sliceCopy)
+	}
+	return newSlices, oldPoolNames, nil
+}
```
+1, this should be ok.

BigDarkClown commented on cluster-autoscaler/simulator/dynamicresources/utils/sanitize.go:
```diff
@@ -0,0 +1,166 @@
+		claimCopy := claim.DeepCopy()
+		claimCopy.UID = uuid.NewUUID()
+		claimCopy.Name = fmt.Sprintf("%s-%s", claim.Name, nameSuffix)
+		claimCopy.OwnerReferences = []metav1.OwnerReference{PodClaimOwnerReference(newOwner)}
```
Ok, makes sense to me.

BigDarkClown commented on cluster-autoscaler/simulator/utilization/info.go:
```diff
@@ -59,6 +60,16 @@ func Calculate(nodeInfo *framework.NodeInfo, skipDaemonSetPods, skipMirrorPods b
+		// TODO(DRA): Figure out how to calculate Node utilization for DRA resources properly. The current logic is provisional, it should work well for Nodes with a single Pool of
+		// expensive Devices but is probably not flexible enough for other scenarios.
+		resourceName, highestUtil, err := drautils.HighestDynamicResourceUtilization(nodeInfo)
+		if err != nil {
```
I think it makes sense, my worry is just that any error in DRA integration that passes through here would de-facto break scale-down. The mitigation in this case is not clear to me - disabling DRA flag seems like the most obvious one, but then we would de facto just fallback to cpu + mem either way.

Overall, I don't want to block on this, especially as the whole utilization of slices is very iffy and provisional. Let's revisit this in the future, once we have a better idea on how this will be handled.

towca replied:
> I think it makes sense, my worry is just that any error in DRA integration that passes through here would de-facto break scale-down.

That would be true only for Nodes using DRA (i.e. exposing ResourceSlices), right? And there would be a mitigation as you mention. IMO this is fine, especially for the MVP. 

> Overall, I don't want to block on this, especially as the whole utilization of slices is very iffy and provisional. Let's revisit this in the future, once we have a better idea on how this will be handled.

Ack!

BigDarkClown commented on cluster-autoscaler/simulator/clustersnapshot/predicate/predicate_snapshot.go:
```diff
@@ -101,6 +170,30 @@ func (s *PredicateSnapshot) SchedulePodOnAnyNodeMatching(pod *apiv1.Pod, anyNode
 
 // UnschedulePod removes the given Pod from the given Node inside the snapshot.
 func (s *PredicateSnapshot) UnschedulePod(namespace string, podName string, nodeName string) error {
+	if s.draEnabled {
```
Okay, the argument for having a single source of implementation for Basic and Delta is good enough for me. I totally forgot during review that we will support both, this is a very nice forward thinking from your side.

BigDarkClown commented on cluster-autoscaler/simulator/clustersnapshot/predicate/predicate_snapshot_test.go:
```diff
@@ -161,58 +355,860 @@ func validTestCases(t *testing.T) []modificationTestCase {
+				// The pod-owned claim already exists in the DRA snapshot, so trying to add it again should be an error.
+				wantErr: cmpopts.AnyError,
+				// The state shouldn't change on error.
+				// TODO(DRA): Until transaction-like clean-up is implemented in AddNodeInfo, the state is not cleaned up on error. Make modifiedState identical to initial state after the clean-up is implemented.
```
Is lack of "transactions" not a blocker for MVP? I am a bit worried that this might easily result in a broken state. Could this be straightforward to implement by adding defer RemoveNodeInfo?

BigDarkClown commented on cluster-autoscaler/simulator/clustersnapshot/predicate/predicate_snapshot_test.go:
```diff
@@ -33,6 +33,8 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
+// TODO(DRA): Add DRA-specific test cases.
```
I think I aged 2 years while going through those, but that looks great now. Thanks!

BigDarkClown reviewed (commented):
Overall this LGTM, thanks for adding **a lot** of tests, I feel much safer now. The only real worry I have now is the AddNodeInfo not being transaction-like. Are you certain this is okay?

pohly commented on cluster-autoscaler/simulator/dynamicresources/utils/sanitize.go:
```diff
@@ -0,0 +1,166 @@
+// SanitizePodResourceClaims can be used to duplicate ResourceClaims needed by a Pod, when duplicating the Pod.
+//   - ResourceClaims owned by oldOwner are duplicated and sanitized, to be owned by a duplicate pod - newOwner.
+//   - ResourceClaims not owned by oldOwner are returned unchanged in the result. They are shared claims not bound to the
+//     lifecycle of the duplicated pod, so they shouldn't be duplicated.
```
> Does this sound correct to you?

I won't pretend that I fully understand the CA logic here, but I trust your judgment, so yes. :sweat_smile: 

> is the max length actually checked in filters?

No. Normally, a workload is set up so that it never exceeds that limit. If that happens, it gets detected by the apiserver when the scheduler plugin tries to update the ResourceClaim, so there is no need to duplicate those checks in the plugin.

In CA, you don't send anything to the apiserver for validation, so you have to replicate that in the code which tracks state.

pohly commented:
Looks like all of my comments have been addressed, so feel free to go ahead as far as I am concerned.

towca commented:
> The only real worry I have now is the AddNodeInfo not being transaction-like. Are you certain this is okay?

All `AddNodeInfo()` calls either break the whole CA loop on error, or are guarded with a `Fork()/Revert()` that guarantees clean-up on error. The same is true for `SchedulePod()/SchedulePodOnAnyNodeMatching()`, with one exception - ProvisioningRequest logic ignores `TrySchedulePods()` errors caused by `SchedulePod()/SchedulePodOnAnyNodeMatching()`. But ProvisioningRequests are not compatible with DRA anyway.

Given that this is all behind an experimental feature flag, IMO this should be fairly okay until the next release. WDYT @BigDarkClown?

towca commented:
Thanks for the quick reviews everyone! I believe this is the current status:

* @pohly LGTMd
* @BigDarkClown The only blocking comment is the one about transcations in `AddNodeInfo()`. Answered above.
* @jackfrancis The only blocking comment is the one about feature gates/flags. Answered there!

BigDarkClown commented:
> > The only real worry I have now is the AddNodeInfo not being transaction-like. Are you certain this is okay?
> 
> All `AddNodeInfo()` calls either break the whole CA loop on error, or are guarded with a `Fork()/Revert()` that guarantees clean-up on error. The same is true for `SchedulePod()/SchedulePodOnAnyNodeMatching()`, with one exception - ProvisioningRequest logic ignores `TrySchedulePods()` errors caused by `SchedulePod()/SchedulePodOnAnyNodeMatching()`. But ProvisioningRequests are not compatible with DRA anyway.
> 
> Given that this is all behind an experimental feature flag, IMO this should be fairly okay until the next release. WDYT @BigDarkClown?

Okay, I feel more confident with that. Thanks for analysis, LGTM from my side, waiting for @jackfrancis to respond and unhold.

/lgtm
/hold

jackfrancis commented:
Great work @towca!

/lgtm

towca commented:
Thanks for the reviews again, merging!

/unhold

sftim commented:
Suggestion for post MVP work: document how to autoscale a cluster that uses dynamically-allocated resources.

ttsuuubasa commented:
Suggestion for post MVP work: Making --gpu-total function work with Dynamic Resource Allocation (DRA).

In my understanding, --gpu-tolal option relies on the counting methodology depending on Device Plugins.
Therefore, this option doesn't work with --enable-dynamic-resource-allocation.
I think ClusterAutoscaler(CA) needs to ask DRA for the way to identify a device in ResourceSlice are GPU in a standard manner.
A field in ResourceSlice to classify a device type currently entrusts to the vendor's free format in "attributes" fields.
Consequently, CA doesn't have a certain method to count GPUs.
Are you willing to support this option when DRA is enabled?

pohly commented:
I don't know what `--gpu-total` does, but a bespoke option specifically for "GPUs" (whatever they are) sounds wrong to me.

Instead of special-case options for specific categories of hardware perhaps the concept can be generalized?

pohly commented:
Categorizing devices doesn't make sense to me and I also find the concept of "counting GPUs" questionable. GPUs can be large and small. A single parameter which applies equally to everything that is a GPU seems like an oversimplification.

But I don't have a stake in this - just my two cents.

jackfrancis commented:
I agree with @pohly on the larger point. I also did some cursory digging and this feature seems to be very incompletely implemented. Documentation suggests it only works with GKE.

My instinct would be to work on deprecating this feature.

What do folks think? cc @MaciekPytel @towca

ttsuuubasa commented:
If this function is deprecated, I will accept it.

> Categorizing devices doesn't make sense to me

  My idea was to add another field to ResourceSlice like spec.devices.type.

> GPUs can be large and small. A single parameter which applies equally to everything that is a GPU seems like an oversimplification.

  This option can specify a gpu type as `--gpu-total=<gpu-type>:<min>:<max>`.
  `<gpu-type>` must match the value of the label in the node group that manages the GPU-enabled nodes of that type and users can set any value.
  I thought this specification can prevent the oversimplification you mentioned.

> this feature seems to be very incompletely implemented. Documentation suggests it only works with GKE.

  It may be needless to say, but I think the document's information is not the most recent and
  this option can be used with Cluster-api and has been adopted in Red Hat OpenShift.

pohly commented:
> My idea was to add another field to ResourceSlice like spec.devices.type.

That is what doesn't make sense to me. Someone would have to define types in such a way that different vendors and users can agree on the definition and then use it consistently. "I know a GPU when I see it" is not good enough.

I suspect the usage of the `--gpu-total` option is to ensure a certain minimum amount of compute power (min value) and to ensure that total costs are not exceeded (max value). In both cases, counting GPUs is only a coarse approximation.

ttsuuubasa commented:
> Someone would have to define types in such a way that different vendors and users can agree on the definition and then use it consistently.

It sounds difficult...
In that case, what do you think about how to solve the issue of "Calculating Node utilization for DRA resources" as written in the description?
It needs to differentiate between DRA devices, whether they are more expensive than CPU/memory or not and whether they should be prioritized or not.
I thought that categorizing the devices as I mentioned would lead to solving this issue.

> I suspect the usage of the --gpu-total option is to ensure a certain minimum amount of compute power (min value) and to ensure that total costs are not exceeded (max value). In both cases, counting GPUs is only a coarse approximation.

I'm on the same page.
